--- common/lib/modules/fglrx/build_mod/firegl_public.c	2015-06-02 16:55:27.252007516 +0300
+++ common/lib/modules/fglrx/build_mod/firegl_public.c	2015-06-02 16:56:15.160008062 +0300
@@ -269,7 +269,7 @@
 #endif
 
 #ifdef MODULE_LICENSE
-MODULE_LICENSE("Proprietary. (C) 2002 - ATI Technologies, Starnberg, GERMANY");
+MODULE_LICENSE("GPL\0Proprietary. (C) 2002 - ATI Technologies, Starnberg, GERMANY");
 #endif
 #ifdef MODULE_DEVICE_TABLE
 MODULE_DEVICE_TABLE(pci, fglrx_pci_table);
