--- /home/me/fdev/catalyst/xorg-x11-drv-catalyst.spec	2013-08-03 15:17:19.000000000 +0300
+++ /home/me/rpmbuild/SOURCES/xorg-x11-drv-catalyst.spec	2013-08-04 18:47:57.092969211 +0300
@@ -21,6 +21,7 @@
 Source8:         00-catalyst-modulepath.conf
 Source9:         01-catalyst-videodriver.conf
 Source10:        blacklist-radeon.conf
+Source11:        http://developer.amd.com/wordpress/media/2012/10/gles_sdk.zip
 
 BuildRoot:       %{_tmppath}/%{name}-%{version}-%{release}-root-%(%{__id_u} -n)
 
@@ -124,6 +125,9 @@
 mv -f README AMD_XvBA_README
 popd
 
+#Extract the GLES SDK
+unzip %{SOURCE11}
+
 # Create tarball of kmod data for use later
 tar -cjf catalyst-kmod-data-%{version}.tar.bz2 fglrx/common/usr/share/doc/fglrx/LICENSE.TXT \
                                             fglrx/common/*/modules/fglrx/ \
@@ -227,6 +231,16 @@
 # Install XvBA headers
 install -D -p -m 0644 amdxvba/include/amdxvba.h $RPM_BUILD_ROOT%{_includedir}/fglrx
 
+# Install GLES libs
+%ifarch %{ix86}
+install -D -p -m 0644 gles_sdk/x86/libEGL.so $RPM_BUILD_ROOT/%{atilibdir}/
+install -D -p -m 0644 gles_sdk/x86/libGLESv2.so $RPM_BUILD_ROOT/%{atilibdir}/
+%endif
+%ifarch x86_64
+install -D -p -m 0644 gles_sdk/x86-64/libEGL.so $RPM_BUILD_ROOT/%{atilibdir}/
+install -D -p -m 0644 gles_sdk/x86/libGLESv2.so $RPM_BUILD_ROOT/%{atilibdir}/
+%endif
+
 # Remove switching scripts
 rm -f $RPM_BUILD_ROOT%{atilibdir}/switchlibGL
 rm -f $RPM_BUILD_ROOT%{atilibdir}/switchlibglx
@@ -259,6 +273,7 @@
 ln -s libXvBAW.so.1.0 $RPM_BUILD_ROOT%{atilibdir}/libXvBAW.so
 ln -s libatiuki.so.1.0 $RPM_BUILD_ROOT%{atilibdir}/libatiuki.so.1
 ln -s libatiuki.so.1.0 $RPM_BUILD_ROOT%{atilibdir}/libatiuki.so
+ln -s libEGL.so $RPM_BUILD_ROOT%{atilibdir}/libEGL.so.1
 
 install -D -p -m 0644 fglrxpkg/usr/share/icons/ccc_large.xpm $RPM_BUILD_ROOT/%{_datadir}/icons/ccc_large.xpm
 install -D -p -m 0755 %{SOURCE4} $RPM_BUILD_ROOT%{_initrddir}/atieventsd
