diff -Naur vdr-yacoto-0.3.0_bak/Makefile vdr-yacoto-0.3.0/Makefile
--- vdr-yacoto-0.3.0_bak/Makefile	2008-10-06 20:14:20.000000000 +0200
+++ vdr-yacoto-0.3.0/Makefile	2010-07-15 01:18:11.250018943 +0200
@@ -22,7 +22,10 @@
 TMPDIR = /tmp
 
 ### The directory containing the conf files
-YAC_CONF_DIR = /etc/vdr/plugins/yacoto
+YAC_CONF_DIR ?= /etc/vdr/plugins/$(PLUGIN)
+
+### The directory containing the script files
+YAC_DIR ?= $(YAC_CONF_DIR)
 
 ### Allow user defined options to overwrite defaults:
 -include $(VDRDIR)/Make.config
@@ -39,7 +42,7 @@
 
 ### Includes and Defines (add further entries here):
 INCLUDES += -I$(VDRDIR)/include
-DEFINES += -D_GNU_SOURCE -DPLUGIN_NAME_I18N='"$(PLUGIN)"'
+DEFINES += -D_GNU_SOURCE -DPLUGIN_NAME_I18N='"$(PLUGIN)"' -DYAC_DIR='"$(YAC_DIR)"'
 
 ### The object files (add further files here):
 OBJS = $(PLUGIN).o yac-menurecordings.o yac-menurecinfo.o yac-menuconvert.o yac-recordings.o yac-tools.o yac-setup.o
diff -Naur vdr-yacoto-0.3.0_bak/README vdr-yacoto-0.3.0/README
--- vdr-yacoto-0.3.0_bak/README	2010-04-08 21:19:24.000000000 +0200
+++ vdr-yacoto-0.3.0/README	2010-07-15 01:24:49.816647011 +0200
@@ -27,7 +27,8 @@
 - Das Verzeichnis in welchem die Scripts stehen (/etc/vdr/plugins/yacoto per default) kann in der
   Make.config mit der Variablen YAC_CONF_DIR geaendert werden.
 
-- die Datei get_vcodec.sh muss auch nach YAC_CONF_DIR kopiert werden.
+- die Datei get_vcodec.sh muss auch nach YAC_DIR kopiert werden, als yac_get_vcodec.sh falls nicht
+  schon vorhanden, aus dem Skript-Pack..
 
 - Anweisungen des Script Paketes befolgen !  
 
diff -Naur vdr-yacoto-0.3.0_bak/yac-menuconvert.c vdr-yacoto-0.3.0/yac-menuconvert.c
--- vdr-yacoto-0.3.0_bak/yac-menuconvert.c	2008-12-18 19:47:40.000000000 +0100
+++ vdr-yacoto-0.3.0/yac-menuconvert.c	2010-07-15 01:18:11.250018943 +0200
@@ -182,7 +182,7 @@
 :cOsdMenu(tr("Convert recordings"), 30)
 {
    if( activeConfig[0] == '\0' ) {
-      cString configFile = cString::sprintf( "%s/yacoto.conf", cPlugin::ConfigDirectory("yacoto"));
+      cString configFile = cString::sprintf( "%s/%s.conf", cPlugin::ConfigDirectory(PLUGIN_NAME_I18N), PLUGIN_NAME_I18N);
 
       FILE *fhdl = fopen( (const char *)configFile, "r" );
       if( fhdl != NULL ) {
diff -Naur vdr-yacoto-0.3.0_bak/yac-recordings.c vdr-yacoto-0.3.0/yac-recordings.c
--- vdr-yacoto-0.3.0_bak/yac-recordings.c	2010-04-08 21:13:45.000000000 +0200
+++ vdr-yacoto-0.3.0/yac-recordings.c	2010-07-15 01:29:16.079535503 +0200
@@ -11,7 +11,7 @@
 #include "yac-recordings.h"
 
 #define INDEXFILESUFFIX   "/index"
-#define SCRIPT_FILE       "yac_start.sh"
+#define SCRIPT_FILE       "/yac_start.sh"
 
 struct tIndexTs {
   uint64_t offset:40; // up to 1TB per file (not using off_t here - must definitely be exactly 64 bit!)
@@ -100,8 +100,8 @@
 const char *cYacRecItem::VideoInfo(void) { 
    if( !videoInfo[0] || !strstr( videoInfo, "CODEC:" )) {
       char commandResult[32];
-      cString cmd = cString::sprintf("%s%s \"%s\"", (const char *)AddDirectory( cPlugin::ConfigDirectory(), "yacoto/" ),
-                                                    "get_vcodec.sh", cRec->FileName());
+      cString cmd = cString::sprintf("%s%s \"%s\"", (const char *)AddDirectory( YAC_DIR, "/" ),
+                                                    "yac_get_vcodec.sh", cRec->FileName());
 
       dsyslog("executing command '%s'", (const char *)cmd);
       FILE *p = popen((const char *)cmd, "r");
@@ -207,7 +207,7 @@
 {
    struct stat statBuf;
 
-   if (lstat(cString::sprintf( "%s/yacoto.log", filename ), &statBuf) == -1) 
+   if (lstat(cString::sprintf( "%s/%s.log", filename, PLUGIN_NAME_I18N ), &statBuf) == -1) 
       return false;
 
    return true;
@@ -253,7 +253,7 @@
    char execCmd[8192];
    int baseLen = 0;
 
-   strcpy(execCmd,AddDirectory( cPlugin::ConfigDirectory(), "yacoto/" ));
+   strcpy(execCmd,YAC_DIR);
    strcat(execCmd,SCRIPT_FILE);
    baseLen = strlen(execCmd);
    if( yacMergeRecs == true || numItems == 1 ) {
diff -Naur vdr-yacoto-0.3.0_bak/yac-setup.c vdr-yacoto-0.3.0/yac-setup.c
--- vdr-yacoto-0.3.0_bak/yac-setup.c	2008-10-06 19:00:00.000000000 +0200
+++ vdr-yacoto-0.3.0/yac-setup.c	2010-07-15 01:18:11.253356236 +0200
@@ -24,7 +24,7 @@
 #define NUM_CHARS   "0123456789"
 #define FNAME_CHARS " abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-.#~,/_@"
 
-#define CFG_FILE "yacadmin.conf"
+#define CFG_FILE "/yacadmin.conf"
 #define START_SCRIPT "/yac_admin.sh -start"
 #define INIT_SCRIPT "/yac_admin.sh -init"
 #define SAVE_SCRIPT "/yac_admin.sh -save"
@@ -45,9 +45,9 @@
    idxConfigVar = -1;
    bChanged = false;
    bStored = false;
-   strcpy(configFile,AddDirectory( cPlugin::ConfigDirectory(), "yacoto/" ));
+   strcpy(configFile,AddDirectory( cPlugin::ConfigDirectory(PLUGIN_NAME_I18N), "/" ));
    strcat(configFile,CFG_FILE);
-   strcpy(buffer,AddDirectory( cPlugin::ConfigDirectory(), "yacoto" ));
+   strcpy(buffer,YAC_DIR);
 
    strcat(buffer, START_SCRIPT);
    dsyslog( "Executing <%s>", buffer );
@@ -262,7 +262,7 @@
    }
    if( bStored ) {
       char buffer[ 512 ];
-      strcpy(buffer,AddDirectory( cPlugin::ConfigDirectory(), "yacoto" ));
+      strcpy(buffer,YAC_DIR);
       strcat(buffer, SAVE_SCRIPT);
       dsyslog( "Executing <%s>", buffer );
       system( buffer );
diff -Naur vdr-yacoto-0.3.0_bak/yac-tools.c vdr-yacoto-0.3.0/yac-tools.c
--- vdr-yacoto-0.3.0_bak/yac-tools.c	2008-10-06 19:00:00.000000000 +0200
+++ vdr-yacoto-0.3.0/yac-tools.c	2010-07-15 01:18:11.318044256 +0200
@@ -17,8 +17,8 @@
       free(commandResult);
       commandResult = NULL;
    }
-   strcpy(cmd,AddDirectory( cPlugin::ConfigDirectory(), "yacoto/" ));
-   strcat(cmd,"yac_show_status.sh");
+   strcpy(cmd,YAC_DIR);
+   strcat(cmd,"/yac_show_status.sh");
 
    dsyslog("executing command '%s'", cmd);
    FILE *p = popen(cmd, "r");
