diff --git a/include/media/saa7146_vv.h b/include/media/saa7146_vv.h


index b9da1f5..4aeff96 100644 (file)


--- a/include/media/saa7146_vv.h

+++ b/include/media/saa7146_vv.h

@@ -188,7 +188,6 @@ void saa7146_buffer_timeout(unsigned long data);

 void saa7146_dma_free(struct saa7146_dev* dev,struct videobuf_queue *q,

                                                struct saa7146_buf *buf);

 

-int saa7146_vv_devinit(struct saa7146_dev *dev);

 int saa7146_vv_init(struct saa7146_dev* dev, struct saa7146_ext_vv *ext_vv);

 int saa7146_vv_release(struct saa7146_dev* dev);

 
