diff --git a/drivers/media/video/hexium_orion.c b/drivers/media/video/hexium_orion.c


index fe596a1..938a1f8 100644 (file)


--- a/drivers/media/video/hexium_orion.c

+++ b/drivers/media/video/hexium_orion.c

@@ -216,10 +216,6 @@ static int hexium_probe(struct saa7146_dev *dev)

                return -EFAULT;

        }

 

-       err = saa7146_vv_devinit(dev);

-       if (err)

-               return err;

-

        hexium = kzalloc(sizeof(struct hexium), GFP_KERNEL);

        if (NULL == hexium) {

                printk("hexium_orion: hexium_probe: not enough kernel memory.\n");
