--- configure.ac.orig	2011-10-09 10:13:40.000000000 -0300
+++ configure.ac	2011-10-19 16:23:44.126656081 -0300
@@ -208,28 +208,28 @@
 if test "x$CXX" = "x" ; then
     have_utp="no"
 else
-    have_utp="yes"
+    have_utp="no"
 fi
 AC_ARG_ENABLE([utp],
               AS_HELP_STRING([--enable-utp],[build ÂµTP support]),
               [want_utp=${enableval}],
               [want_utp=${have_utp}])
-if test "x$want_utp" = "xyes" ; then
-    if test "x$have_utp" = "xyes"; then 
+if test "x$want_utp" = "xno" ; then
+    if test "x$have_utp" = "xno"; then 
         LIBUTP_CFLAGS="-I\$(top_srcdir)/third-party/"
         LIBUTP_LIBS="\$(top_builddir)/third-party/libutp/libutp.a"
 	if test "x$libutp_extra_libs" != "x" ; then
 	   LIBUTP_LIBS="$LIBUTP_LIBS $libutp_extra_libs"
 	fi
         AC_DEFINE([WITH_UTP],[1])
-        build_utp="yes"
+        build_utp="no"
     else 
       AC_MSG_ERROR("Unable to build uTP support -- C++ compiler not found") 
     fi 
 fi 
 AC_SUBST(LIBUTP_CFLAGS)
 AC_SUBST(LIBUTP_LIBS)
-AM_CONDITIONAL([BUILD_UTP],[test "x$build_utp" = "xyes"])
+AM_CONDITIONAL([BUILD_UTP],[test "x$build_utp" = "xno"])
 AC_MSG_RESULT([$build_utp])
