--- setup.py.orig	2010-02-23 11:23:54.792669992 +0100
+++ setup.py	2010-02-23 11:27:41.868506250 +0100
@@ -118,6 +118,7 @@
     data_files = list_from_lists(
         [('share/gnome-activity-journal', ['gnome-activity-journal'])],
         [('share/gnome-activity-journal/data', glob('data/*.svg'))],
+        [('share/gnome-activity-journal/data/zlogo', glob('data/zlogo/*.png'))],
         [('share/pixmaps/', ['data/gnome-activity-journal.xpm'])],
         recursive_install('share/icons/hicolor', 'data/icons/hicolor/', '',
             ext=['.png', '.svg']),
