Index: gnome-activity-journal-0.3.3/setup.py
===================================================================
--- gnome-activity-journal-0.3.3.orig/setup.py
+++ gnome-activity-journal-0.3.3/setup.py
@@ -55,7 +55,7 @@ class _install(orig_install):
             'bin/gnome-activity-journal')
         # Symlink the 48x48 PNG icon into share/pixmaps
         self._create_symlink(
-            'share/gnome-activity-journal/data/icons/hicolor/48x48/apps/' \
+            'share/icons/hicolor/48x48/apps/' \
                 'gnome-activity-journal.png',
             'share/pixmaps/gnome-activity-journal.png')
         # Symlink the icons so that the Journal can find them
