--- a/include/the_Foundation/c11threads.h
+++ b/include/the_Foundation/c11threads.h
@@ -31,7 +31,10 @@
 typedef pthread_mutex_t mtx_t;
 typedef pthread_cond_t cnd_t;
 typedef pthread_key_t tss_t;
+#ifndef __once_flag_defined
+#define ONCE_FLAG_INIT PTHREAD_ONCE_INIT
 typedef pthread_once_t once_flag;
+#endif
 
 typedef int (*thrd_start_t)(void*);
 typedef void (*tss_dtor_t)(void*);
@@ -90,7 +93,9 @@
 
 /* ---- misc ---- */
 
+#ifndef __once_flag_defined
 void call_once(once_flag *flag, void (*func)(void));
+#endif
 
 iEndPublic
 
--- a/src/string.c
+++ b/src/string.c
@@ -1198,5 +1198,5 @@
 static iThreadStorage *threadLocalCharBuffer_(void) {
-    static once_flag inited = PTHREAD_ONCE_INIT;
+    static once_flag inited = ONCE_FLAG_INIT;
     call_once(&inited, initThreadLocalCharBuffer_);
     return threadLocalCharBuffer_;
 }
