--- a/include/the_Foundation/c11threads.h
+++ b/include/the_Foundation/c11threads.h
@@ -31,7 +31,9 @@
 #include <pthread.h>

-#define ONCE_FLAG_INIT PTHREAD_ONCE_INIT
+#ifndef __once_flag_defined
+#define ONCE_FLAG_INIT PTHREAD_ONCE_INIT
+#endif

 typedef pthread_cond_t cnd_t;
 typedef pthread_t thrd_t;
@@ -31,7 +33,9 @@
 typedef pthread_key_t tss_t;
 typedef pthread_mutex_t mtx_t;
+#ifndef __once_flag_defined
 typedef pthread_once_t once_flag;
+#endif

 /* ... */

+#ifndef __once_flag_defined
 void call_once(once_flag *flag, void (*func)(void));
+#endif
