From e25e2951f04ac6ca5ba7d4ac051f59eafebddfdd Mon Sep 17 00:00:00 2001
From: Fabian Deutsch <fabian.deutsch@gmx.de>
Date: Wed, 6 Apr 2011 14:55:44 +0200
Subject: [PATCH 1/2] Add GConfSwitchTweak for boolean gconf values.

Signed-off-by: Fabian Deutsch <fabian.deutsch@gmx.de>
---
 gtweak/widgets.py |   13 +++++++++++++
 1 files changed, 13 insertions(+), 0 deletions(-)

diff --git a/gtweak/widgets.py b/gtweak/widgets.py
index 110cfce..2e88ea2 100644
--- a/gtweak/widgets.py
+++ b/gtweak/widgets.py
@@ -157,6 +157,19 @@ class _GConfTweak(Tweak):
             self.gconf.schema_get_description(),
             **options)
 
+class GConfSwitchTweak(_GConfTweak):
+    def __init__(self, key_name, **options):
+        _GConfTweak.__init__(self, key_name, bool, **options)
+
+        w = Gtk.Switch()
+	w.set_active(self.gconf.get_value())
+        w.connect('notify::active', self._on_w_changed)
+        self.widget = build_label_beside_widget(self.name, w)
+        self.widget_for_size_group = w
+
+    def _on_w_changed(self, w, v):
+        self.gconf.set_value(w.get_active())
+
 class GConfComboTweak(_GConfTweak):
     def __init__(self, key_name, key_type, key_options, **options):
         _GConfTweak.__init__(self, key_name, key_type, **options)
-- 
1.7.4.2

